projects/common/src/main/java/dan200/computercraft/shared/turtle/core/TurtlePlaceCommand.java

```diff
@@ -245,8 +310,150 @@ private static void setSignText(Level world, SignBlockEntity sign, String messag
+            else if (orientationParams.blockFacing.getAxis() != Direction.Axis.Y && newState.hasProperty(BlockStateProperties.HORIZONTAL_FACING)) {
+                newState = newState.setValue(BlockStateProperties.HORIZONTAL_FACING, orientationParams.blockFacing);
+            }
+            // Try AXIS for logs and pillars
+            else if (newState.hasProperty(BlockStateProperties.AXIS)) {
+                newState = newState.setValue(BlockStateProperties.AXIS, orientationParams.blockFacing.getAxis());
+            }
+        }
+
+        // Handle slab type (top/bottom)
+        if (orientationParams.isTop != null && newState.hasProperty(BlockStateProperties.SLAB_TYPE)) {
+            newState = newState.setValue(BlockStateProperties.SLAB_TYPE, orientationParams.isTop ? SlabType.TOP : SlabType.BOTTOM);
+        }
+
+        // Handle stair half (top/bottom for upside-down)
+        if (orientationParams.isUpsideDown && newState.hasProperty(BlockStateProperties.HALF)) {
+            newState = newState.setValue(BlockStateProperties.HALF, Half.TOP);
+        }
+
+        return newState;
+    }
+
     private static final class ErrorMessage {
         @Nullable
         String message;
     }
+
+    private static final class OrientationParameters {
+        @Nullable Direction blockFacing;
+        @Nullable Boolean isTop; // null means no slab preference, true = top, false = bottom
+        boolean isUpsideDown;
+        boolean isGroundAttachment; // for torches on ground vs wall
+        boolean foundOrientationParam;
+        @Nullable Direction clickFace; // Which face to click on target block
+
+        OrientationParameters(@Nullable Direction blockFacing, @Nullable Boolean isTop, boolean isUpsideDown, boolean isGroundAttachment, boolean foundOrientationParam, @Nullable Direction clickFace) {
+            this.blockFacing = blockFacing;
+            this.isTop = isTop;
+            this.isUpsideDown = isUpsideDown;
+            this.isGroundAttachment = isGroundAttachment;
+            this.foundOrientationParam = foundOrientationParam;
+            this.clickFace = clickFace;
+        }
+    }
+
+    private static OrientationParameters parseOrientationParameters(@Nullable Object[] extraArguments, Direction turtleFacing) {
+        Direction blockFacing = null;
+        Boolean isTop = null; // null means no slab preference, true = top, false = bottom
+        boolean isUpsideDown = false;
+        boolean isGroundAttachment = false; // for torches on ground vs wall
+        boolean foundOrientationParam = false;
+        Direction clickFace = null;
+
+        if (extraArguments != null && extraArguments.length > 0) {
+            // Process all string arguments as potential orientation parameters
+            for (var arg : extraArguments) {
+                if (arg instanceof String orientation) {
+                    switch (orientation.toLowerCase()) {
```

Comment:
We should use the ROOT locale, to avoid weird system-specific issues.

```suggestion
                    switch (orientation.toLowerCase(Locale.ROOT)) {
```

I don't know if it's also worth checking unknown orientations, and rejecting invalid combinations (e.g. `"left", "right"`)?